lasagne/layers/sparse/noise.py

```diff
@@ -0,0 +1,40 @@
+import theano
+import theano.sparse as sparse
+
+from ..noise import DropoutLayer as Dropout
+
+
+__all__ = [
+    "DropoutLayer",
+    "dropout",
+]
+
+
+class DropoutLayer(Dropout):
+    def get_output_for(self, input, deterministic=False, **kwargs):
+        """
+        Parameters
+        ----------
+        input : theano sparse matrix
+            output from the previous layer
+        deterministic : bool
+            If true dropout and scaling is disabled, see notes
+        """
+        if deterministic or self.p == 0:
+            return input
+        else:
+            assert type(input) == sparse.basic.SparseVariable
+
+            retain_prob = 1 - self.p
+            if self.rescale:
+                input = sparse.basic.mul(input, 1/retain_prob)
```

Comment:
Weird! Seems it makes a difference whether `retain_prob` is a numpy scalar or a Python float, but in theory it shouldn't. Maybe Theano passes is on to scipy as-is, and scipy uses a different algorithm that needlessly uses multi-threading.
In this case, we need to check whether `self.p` is symbolic or not, and use `float(1 - self.p)` if it is not symbolic:
```python
if isinstance(self.p, theano.Variable):
    retain_prob = 1 - self.p
    inv_retain_prob = T.inv(self.p)
else:
    retrain_prob = float(1 - self.p)
    inv_retain_prob = 1 / retrain_prob
if self.rescale:
    input = sparse.basic.mul(input, inv_retain_prob)
```